src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+  }
+
+  ~C_WriteRequest() {
+    ldout(m_cct, 99) << this << dendl;
+  }
+
+  void complete_user_request(int r) {
+    bool initial = false;
+    if (m_user_req_completed.compare_exchange_strong(initial, true)) {
+      ldout(m_cct, 15) << this << " completing user req" << dendl;
+      user_req->complete(r);
+    } else {
+      ldout(m_cct, 20) << this << " user req already completed" << dendl;
+    }
+  }
+
+  virtual void send() override {
+    /* Should never be called */
+    ldout(m_cct, 6) << this << " unexpected" << dendl;
+  }
+  
+  virtual void finish(int r) {
+    ldout(m_cct, 6) << this << dendl;
+
+    complete_user_request(r);
+    _on_finish->complete(0);
+  }
+
+  virtual const char *get_name() const override {
+    return "C_WriteRequest";
+  }  
+};
+
+/*
+ * Performs the log event append operation for all of the scheduled
+ * events.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::append_scheduled_ops(void)
+{
+  WriteLogOperations ops;
+  int append_result = 0;
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (m_ops_to_append.empty()) {
+      return;
+    }
+  }
+  {
+    Mutex::Locker locker(m_log_append_lock);
+
+    {
+      Mutex::Locker locker(m_lock);
+      std::swap(ops, m_ops_to_append);
+    }
+    
+    if (ops.size()) {
+      alloc_op_log_entries(ops);
+      append_result = append_op_log_entries(ops);
+    }
+  }
+
+  if (ops.size()) {
+    complete_op_log_entries(ops, append_result);
+  }
+}
+
+/*
+ * Takes custody of ops. They'll all get their log entries appended,
+ * and have their on_write_persist contexts completed once they and
+ * all prior log entries are persisted everywhere.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::schedule_append(WriteLogOperations &ops)
+{
+  {
+    Mutex::Locker locker(m_lock);
+
+    m_ops_to_append.splice(m_ops_to_append.end(), ops);
+  }
+
+  m_log_append_finisher.queue(new FunctionContext([this](int r) {
+	append_scheduled_ops();
+      }));
+}
+
+/*
+ * Performs the pmem buffer flush on all scheduled ops, then schedules
+ * the log event append operation for all of them.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::flush_then_append_scheduled_ops(void)
+{
+  WriteLogOperations ops;
+  {
+    Mutex::Locker locker(m_lock);
+    std::swap(ops, m_ops_to_flush);
+  }
+
+  /* Ops subsequently scheduled for flush may finish before these,
+   * which is fine. We're unconcerned with completion order until we
+   * get to the log message append step. */
+  if (ops.size()) {
+    flush_pmem_buffer(ops);
+    schedule_append(ops);
+  }
+}
+
+/*
+ * Takes custody of ops. They'll all get their pmem blocks flushed,
+ * then get their log entries appended.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::schedule_flush_and_append(WriteLogOperations &ops)
+{
+  {
+    Mutex::Locker locker(m_lock);
+
+    m_ops_to_flush.splice(m_ops_to_flush.end(), ops);
+  }
+
+  m_persist_finisher.queue(new FunctionContext([this](int r) {
```

Comment:
Nit: would you ever need more than a single persist finisher queued concurrently? i.e. if the ```m_ops_top_flush``` queue is non-empty when you append to it, you should know that a callback is already queued.

Reply:
We don't need >1 finisher queued.